pkg/apis/proto/source/v1/udsource.proto

```diff
@@ -0,0 +1,138 @@
+syntax = "proto3";
+
+option go_package = "github.com/numaproj/numaflow-go/pkg/apis/proto/source/v1";
+
+import "google/protobuf/timestamp.proto";
+import "google/protobuf/empty.proto";
+
+package source.v1;
+
+service UserDefinedSource {
+  // Read returns a stream of datum responses.
+  // The size of the returned ReadResponse is less than or equal to the num_records specified in ReadRequest.
+  rpc Read(ReadRequest) returns (stream ReadResponse);
+
+  // Ack acknowledges a list of datum offsets.
+  // It indicates that the datum stream has been processed by the source vertex.
+  rpc Ack(AckRequest) returns (AckResponse);
+
+  // Pending returns the number of pending records at the user defined source.
+  rpc Pending(google.protobuf.Empty) returns (PendingResponse);
+
+  // IsSourceReady is the heartbeat endpoint for user defined source gRPC.
+  rpc IsSourceReady(google.protobuf.Empty) returns (SourceReadyResponse);
+}
+
+/*
+ * EventTime represents the event time of the datum.
+ */
+message EventTime {
+  google.protobuf.Timestamp event_time = 1;
+}
+
+/*
+ * ReadRequest is the request for reading datum stream from user defined source.
+ */
+message ReadRequest {
+  message Request {
```

Comment:
Do we need to define a struct `ReadRequest` which only contains a single field `Request`?  Why not using a one layer struct? This is a common question for all the similar definitions in this PR.

Reply:
Thanks! I followed the proto3 API guideline https://protobuf.dev/programming-guides/api/#dont-include-primitive-types

Reply:
Good to know!